# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/script/ModuleLoader.h
# Commit: 3a34d557d474
# Full Hash: 3a34d557d474adff2245632b993529dd8e01cb3a
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-02-18 09:08:22
# Regressor Bug: 1742437
# File Overlap Count: 1
# Description:
#   Bug 1742437 - move DOM module loader into own file; r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D137531
# ==============================================================================

diff -r f29451ea18dc -r 3a34d557d474 dom/script/ModuleLoader.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/script/ModuleLoader.h	Thu Feb 17 15:57:51 2022 +0000
@@ -0,0 +1,79 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_dom_ModuleLoader_h
+#define mozilla_dom_ModuleLoader_h
+
+#include "ModuleLoaderBase.h"
+#include "mozilla/dom/ScriptLoadRequest.h"
+#include "ScriptLoader.h"
+
+class nsIURI;
+
+namespace JS {
+
+class CompileOptions;
+
+}  // namespace JS
+
+namespace mozilla {
+namespace dom {
+
+class ModuleLoadRequest;
+class SRIMetadata;
+
+//////////////////////////////////////////////////////////////
+// DOM Module loader implementation
+//////////////////////////////////////////////////////////////
+
+class ModuleLoader final : public ModuleLoaderBase {
+ private:
+  virtual ~ModuleLoader();
+
+ public:
+  explicit ModuleLoader(ScriptLoader* aLoader);
+
+  ScriptLoader* GetScriptLoader();
+
+  // Methods that must be overwritten by an extending class
+  void EnsureModuleHooksInitialized() override;
+
+  /**
+   * Start a load for a module script URI.
+   * Sets up the necessary security flags before calling StartLoadInternal.
+   * Short-circuits if the module is already being loaded.
+   */
+  nsresult StartModuleLoad(ScriptLoadRequest* aRequest) override;
+
+  void ProcessLoadedModuleTree(ModuleLoadRequest* aRequest) override;
+
+  nsresult CompileOrFinishModuleScript(
+      JSContext* aCx, JS::Handle<JSObject*> aGlobal,
+      JS::CompileOptions& aOptions, ModuleLoadRequest* aRequest,
+      JS::MutableHandle<JSObject*> aModuleScript) override;
+
+  // Create a top-level module load request.
+  static already_AddRefed<ModuleLoadRequest> CreateTopLevel(
+      nsIURI* aURI, ScriptFetchOptions* aFetchOptions,
+      const SRIMetadata& aIntegrity, nsIURI* aReferrer, ScriptLoader* aLoader,
+      DOMScriptLoadContext* aContext);
+
+  // Create a module load request for a static module import.
+  already_AddRefed<ModuleLoadRequest> CreateStaticImport(
+      nsIURI* aURI, ModuleLoadRequest* aParent) override;
+
+  // Create a module load request for dynamic module import.
+  static already_AddRefed<ModuleLoadRequest> CreateDynamicImport(
+      nsIURI* aURI, ScriptFetchOptions* aFetchOptions, nsIURI* aBaseURL,
+      DOMScriptLoadContext* aContext, ScriptLoader* aLoader,
+      JS::Handle<JS::Value> aReferencingPrivate,
+      JS::Handle<JSString*> aSpecifier, JS::Handle<JSObject*> aPromise);
+};
+
+}  // namespace dom
+}  // namespace mozilla
+
+#endif  // mozilla_dom_ModuleLoader_h